File: backend/confidence.py
"""
Confidence Meter calculator — computes a weighted "voting readiness" score.

Weights (from design doc):
  critical = 3
  high     = 2
  medium   = 1

confidence_pct = (earned_weight / max_weight) * 100

Returns 0.0 when the step list is empty.
Result is always in [0.0, 100.0].
"""
from __future__ import annotations

from typing import List

from .models import Step, IMPORTANCE_WEIGHTS


def compute_confidence(steps: List[Step]) -> float:
    """
    Return a weighted readiness score as a percentage in [0.0, 100.0].
    Only steps with status 'done' contribute to the earned weight.
    """
    if not steps:
        return 0.0

    max_weight = sum(IMPORTANCE_WEIGHTS.get(s.importance, 1) for s in steps)
    if max_weight == 0:
        return 0.0

    earned = sum(
        IMPORTANCE_WEIGHTS.get(s.importance, 1)
        for s in steps
        if s.status == "done"
    )
    return round((earned / max_weight) * 100, 2)

File: frontend/static/js/confidence.js
/**
 * ConfidenceMeter — computes weighted voting readiness score and updates the UI.
 * Weights: critical=3, high=2, medium=1  (mirrors backend confidence.py)
 * Requirements: Confidence Meter feature
 */

const ConfidenceMeter = (() => {
  const WEIGHTS = { critical: 3, high: 2, medium: 1 };

  let _fill  = null;
  let _value = null;

  function init() {
    _fill  = document.getElementById('confidence-fill');
    _value = document.getElementById('confidence-value');
  }

  /**
   * Update the confidence meter.
   * @param {Array} steps - [{importance, status}, ...]
   */
  function update(steps) {
    if (!steps || steps.length === 0) {
      _set(0);
      return;
    }

    const maxWeight = steps.reduce((sum, s) => sum + (WEIGHTS[s.importance] || 1), 0);
    if (maxWeight === 0) { _set(0); return; }

    const earned = steps
      .filter(s => s.status === 'done')
      .reduce((sum, s) => sum + (WEIGHTS[s.importance] || 1), 0);

    const pct = Math.round((earned / maxWeight) * 100);
    _set(pct);
  }

  function _set(pct) {
    const clamped = Math.min(100, Math.max(0, pct));
    if (_fill)  _fill.style.width    = `${clamped}%`;
    if (_value) _value.textContent   = `${clamped}%`;
  }

  return { init, update };
})();
